core/templates/dev/head/pages/exploration_editor/ExplorationSaveService.js

```diff
@@ -0,0 +1,550 @@
+      focusService, changeListService, siteAnalyticsService) {
+    // Whether or not a save action is currently in progress
+    // (request has been sent to backend but no reply received yet)
+    var saveIsInProgress = false;
+
+    // This flag is used to ensure only one save exploration modal can be open
+    // at any one time.
+    var modalIsOpen = false;
+
+    var diffData;
+
+    var isAdditionalMetadataNeeded = function() {
+      return (
+        !explorationTitleService.savedMemento ||
+        !explorationObjectiveService.savedMemento ||
+        !explorationCategoryService.savedMemento ||
+        explorationLanguageCodeService.savedMemento ===
+          GLOBALS.DEFAULT_LANGUAGE_CODE ||
+        explorationTagsService.savedMemento.length === 0);
+    };
+
+    var areRequiredFieldsFilled = function() {
+      if (!explorationTitleService.displayed) {
+        alertsService.addWarning('Please specify a title');
+        return false;
+      }
+      if (!explorationObjectiveService.displayed) {
+        alertsService.addWarning('Please specify an objective');
+        return false;
+      }
+      if (!explorationCategoryService.displayed) {
+        alertsService.addWarning('Please specify a category');
+        return false;
+      }
+
+      return true;
+    };
+
+    var showCongratulatorySharingModal = function() {
+      return $modal.open({
+        templateUrl: 'modals/shareExplorationAfterPublish',
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance', 'explorationContextService',
+          'UrlInterpolationService',
+          function($scope, $modalInstance, explorationContextService,
+            UrlInterpolationService) {
+            $scope.congratsImgUrl = UrlInterpolationService.getStaticImageUrl(
+              '/general/congrats.svg');
+            $scope.DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR = (
+              GLOBALS.DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR);
+            $scope.close = function() {
+              $modalInstance.dismiss('cancel');
+            };
+            $scope.explorationId = (
+              explorationContextService.getExplorationId());
+          }
+        ]
+      });
+    };
+
+    var openPublishExplorationModal = function(toggleLoadingDotsCallback) {
+      // This is resolved when modal is closed.
+      var deferred = $q.defer();
+
+      var publishModalInstance = $modal.open({
+        templateUrl: 'modals/publishExploration',
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance', function($scope, $modalInstance) {
+            $scope.publish = $modalInstance.close;
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+              deferred.resolve();
+            };
+          }
+        ]
+      });
+
+      publishModalInstance.result.then(function() {
+        // Toggle dots on.
+        if (toggleLoadingDotsCallback) {
```

Comment:
Pass separately-named callbacks for each of the events you care about, here and below.